File: sami/step_definitions/Product Filters.js
describe('Product Filters', () => {
  beforeEach(() => {
    cy.visit('/product-search');
  });

  it('should filter products by price range', () => {
    cy.get('.price-filter').select('20-50');
    cy.get('.search-results').should('contain', 'Price: 20-50');
  });

  it('should filter products by rating', () => {
    cy.get('.rating-filter').select('4 stars');
    cy.get('.search-results').should('contain', 'Rating: 4 stars');
  });

  it('should filter products by multiple criteria', () => {
    cy.get('.price-filter').select('20-50');
    cy.get('.rating-filter').select('4 stars');
    cy.get('.search-results').should('contain', 'Price: 20-50').and('contain', 'Rating: 4 stars');
  });

  it('should clear all filters', () => {
    cy.get('.price-filter').select('20-50');
    cy.get('.rating-filter').select('4 stars');
    cy.get('.clear-filters').click();
    cy.get('.search-results').should('not.contain', 'Price: 20-50').and('not.contain', 'Rating: 4 stars');
  });

  it('should display no products found message when no products match', () => {
    cy.get('.price-filter').select('1000-2000');
    cy.get('.rating-filter').select('5 stars');
    cy.get('.search-results').should('contain', 'No products were found');
  });

  it('should persist filters after page refresh', () => {
    cy.get('.price-filter').select('20-50');
    cy.get('.rating-filter').select('4 stars');
    cy.reload();
    cy.get('.price-filter').should('have.value', '20-50');
    cy.get('.rating-filter').should('have.value', '4 stars');
  });
});

File: step_definitions/Product Filters.java
package com.cucumber.steps;

import com.framework.cucumber.TestHarness;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;

public class ProductFiltersSteps extends SearchResultsPage {

    private TestHarness testHarness = new TestHarness();

    @Given("the user is on the search results page")
    public void theUserIsOnTheSearchResultsPage() {
        String searchResultsURL = testHarness.getData("SearchData", "SearchResultsPageURL");
        navigateToPage(searchResultsURL);
    }

    @When("the user selects a price range filter")
    public void theUserSelectsAPriceRangeFilter() {
        String priceRange = testHarness.getData("FilterData", "PriceRange");
        applyPriceFilter(priceRange);
    }

    @Then("the search results should only display products within the selected price range")
    public void theSearchResultsShouldOnlyDisplayProductsWithinTheSelectedPriceRange() {
        String expectedPriceRange = testHarness.getData("FilterData", "PriceRange");
        verifyProductsWithinPriceRange(expectedPriceRange);
    }

    @When("the user selects a rating filter")
    public void theUserSelectsARatingFilter() {
        String rating = testHarness.getData("FilterData", "Rating");
        applyRatingFilter(rating);
    }

    @Then("the search results should only display products with the selected rating or higher")
    public void theSearchResultsShouldOnlyDisplayProductsWithTheSelectedRatingOrHigher() {
        String expectedRating = testHarness.getData("FilterData", "Rating");
        verifyProductsWithRatingOrHigher(expectedRating);
    }

    @When("the user selects multiple filters such as price and rating")
    public void theUserSelectsMultipleFiltersSuchAsPriceAndRating() {
        String priceRange = testHarness.getData("FilterData", "PriceRange");
        String rating = testHarness.getData("FilterData", "Rating");
        applyMultipleFilters(priceRange, rating);
    }

    @Then("the search results should only display products that match all selected criteria")
    public void theSearchResultsShouldOnlyDisplayProductsThatMatchAllSelectedCriteria() {
        String expectedPriceRange = testHarness.getData("FilterData", "PriceRange");
        String expectedRating = testHarness.getData("FilterData", "Rating");
        verifyProductsMatchAllCriteria(expectedPriceRange, expectedRating);
    }

    @Given("the user has applied filters on the search results page")
    public void theUserHasAppliedFiltersOnTheSearchResultsPage() {
        String priceRange = testHarness.getData("FilterData", "PriceRange");
        String rating = testHarness.getData("FilterData", "Rating");
        applyMultipleFilters(priceRange, rating);
    }

    @When("the user clicks on the clear filters option")
    public void theUserClicksOnTheClearFiltersOption() {
        clearAllFilters();
    }

    @Then("all filters should be removed and the search results should display all products")
    public void allFiltersShouldBeRemovedAndTheSearchResultsShouldDisplayAllProducts() {
        verifyAllProductsDisplayed();
    }

    @Then("the user should see filter options for price, rating, and other criteria")
    public void theUserShouldSeeFilterOptionsForPriceRatingAndOtherCriteria() {
        verifyFilterOptionsVisible();
    }

    @When("the user selects a filter by a specific criterion \(e.g., brand, category\)")
    public void theUserSelectsAFilterByASpecificCriterion() {
        String criterion = testHarness.getData("FilterData", "Criterion");
        applySpecificCriterionFilter(criterion);
    }

    @Then("the search results should only display products that match the selected criterion")
    public void theSearchResultsShouldOnlyDisplayProductsThatMatchTheSelectedCriterion() {
        String expectedCriterion = testHarness.getData("FilterData", "Criterion");
        verifyProductsMatchCriterion(expectedCriterion);
    }
}